### ISSUE TITLE
Entries go missing from autoload_namespaces.php with --no-dev

### ISSUE THREAD

**Issue by glensc** [composer/composer#8485]
The reproducer is in [a gist](https://gist.github.com/glensc/3bcf4c1a4406ac730ea574e3fb2ebc68):

```
git clone -b v1 https://gist.github.com/glensc/3bcf4c1a4406ac730ea574e3fb2ebc68 zf1s-17
cd zf1s-17
composer install --no-dev
php test.php
```

Running composer install vs composer install --no-dev removes following entry from `vendor/composer/autoload_namespaces.php`:

```diff
-    'Zend_Loader' => array($vendorDir . '/zf1s/zend-loader/library'),
```

But it's in normal dependency, not dev dependency.

The workaround is to execute `composer dump` to restore proper autoload namepace mapping.

I believe this has something to do with `"replace"` directives used in `"zf1"` -> `"z1fs"` namespace transition.

You can see more details of this problem in [zf1s](https://github.com/zf1s/zf1) project issue:
- https://github.com/zf1s/zf1/issues/17

**Comment by glensc** [comment]
minimized test case with just 1 dependency in composer.json:

```
git clone -b v2 https://gist.github.com/glensc/3bcf4c1a4406ac730ea574e3fb2ebc68 zf1s-17
cd zf1s-17
composer install --no-dev
php test.php
```

**Comment by stof** [comment]
@glensc the question is how did you got the composer.lock that is in your gist ? If I run a `composer update` with the composer.json file in this gist, I'm **not** getting the zf1s packages that you have in your lock file. That might explain why Composer goes weird when trying to keep only non-dev packages in the autoload.

**Comment by glensc** [comment]
@stof: composer require <new_namespace_package> & commit composer.lock

**Comment by Seldaek** [comment]
The problem is your composer.json is not up to date with the lock file IMO.

If I do a composer update, all the zf1s packages get removed in favor of zf1 ones, then things work fine. If I change the composer.json to require zf1s packages instead of zf1 ones, then `composer dump --no-dev` also starts working fine.

The issue is the no-dev flag walks all packages recursively from the composer.json file, not from the lock file, so if the composer.json does not match lock file you get problems.

**Comment by glensc** [comment]
This maybe fixed by #8621 or needs similar fix.
